src/twisted/application/internet.py

```diff
@@ -522,14 +523,388 @@ def policy(attempt):
+                       endpoint=self._endpoint, delay=delay)
+        self._retryCall = self._clock.callLater(delay, self._reconnect)
+
+
+    @_machine.input()
+    def _reconnect(self):
+        """
+        The wait between connection attempts is done.
+        """
+
+    @_machine.input()
+    def _clientDisconnected(self):
+        """
+        The current connection has been disconnected.
+        """
+
+    @_machine.output()
+    def _forgetConnection(self):
+        """
+        Forget the current connection.
+        """
+        del self._currentConnection
+
+
+    @_machine.output()
+    def _cancelConnectWaiters(self):
+        """
+        Notify all pending requests for a connection that no more connections
+        are expected.
+        """
+        self._unawait(Failure(CancelledError()))
+
+
+    @_machine.output()
+    def _finishStopping(self):
+        """
+        Notify all deferreds waiting on the service stopping.
+        """
+        self._stopWaiters, waiting = [], self._stopWaiters
+        for w in waiting:
+            w.callback(None)
+
+
+    @_machine.input()
+    def whenConnected(self):
+        """
+        Retrieve the currently-connected L{Protocol}, or the next one to
+        connect.
+
+        @return: a Deferred that fires with a protocol produced by the factory
+            passed to C{__init__}
+        @rtype: L{Deferred} firing with L{IProtocol} or failing with
+            L{CancelledError} the service is stopped.
+        """
+
+    @_machine.output()
+    def _currentConnection(self):
+        """
+        Return the currently connected protocol.
+
+        @return: L{Deferred} that is fired with currently connected protocol.
+        """
+        return succeed(self._currentConnection)
+
+
+    @_machine.output()
+    def _noConnection(self):
+        """
+        Notify the caller that no connection is expected.
+
+        @return: L{Deferred} that is fired with L{CancelledError}.
+        """
+        return fail(CancelledError())
+
+
+    @_machine.output()
+    def _awaitingConnection(self):
+        """
+        Return a deferred that will fire with the next connected protocol.
+
+        @return: L{Deferred} that will fire with the next connected protocol.
+        """
+        result = Deferred()
+        self._awaitingConnected.append(result)
+        return result
+
+
+    def _unawait(self, value):
+        """
+        Fire all outstanding L{ClientService.whenConnected} L{Deferred}s.
+
+        @param value: the value to fire the L{Deferred}s with.
+        """
+        self._awaitingConnected, waiting = [], self._awaitingConnected
+        for w in waiting:
+            w.callback(value)
+
+    # State Transitions
+
+    _init.upon(start, enter=_connecting,
+               outputs=[_connect])
+    _init.upon(stop, enter=_init,
+               outputs=[])
+
+    _connecting.upon(start, enter=_connecting, outputs=[])
+    _connecting.upon(stop, enter=_disconnecting,
```

Comment:
I feel like a lot of these `upon` invocations are bloating the line-count by wrapping lines too early.  It's a matter of taste though, so if you prefer this formatting you can leave it.